File: lc/coreprep/implementation/practice_iterations/main_1/NextPermutation.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class NextPermutation {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();

        int[] a = new int[n];
        for (int i = 0; i < n; i++) {
          a[i] = in.nextInt();
        }

        Arrays.sort(a);

        do {
          for (int e : a) {
            System.out.printf("%d ", e);
          }
          System.out.println();
        } while (nextPermutation(a));
      }
    }
  }

  // O(n) / O(1)
  private static boolean nextPermutation(int[] a) {
    int invPos = -1;
    for (int i = a.length - 1; i > 0; i--) {
      if (a[i] > a[i - 1]) {
        invPos = i - 1;
        break;
      }
    }

    if (invPos == -1) {
      return false;
    }

    int nextPos = nextGreater(a, invPos, a.length - 1);
    swap(a, invPos, nextPos);
    reverse(a, invPos + 1, a.length - 1);

    return true;
  }

  private static int nextGreater(int[] a, int invPos, int n) {
    for (int i = n; i > invPos; i--) {
      if (a[i] > a[invPos]) {
        return i;
      }
    }

    return -1;
  }

  private static void swap(int[] a, int x, int y) {
    int t = a[x];
    a[x] = a[y];
    a[y] = t;
  }

  private static void reverse(int[] a, int from, int to) {
    for (int i = from, j = to; i < j; i++, j--) {
      swap(a, i, j);
    }
  }
}

File: lc/coreprep/implementation/practice_iterations/iteration_archives/tidbits_2/next_permutation.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

void swap(vector<int> &a, int x, int y) {
  int t = a[x];
  a[x] = a[y];
  a[y] = t;
}

void reverse(vector<int> &a, int from, int to) {
  for (int i = from, j = to; i < j; i++, j--) {
    swap(a, i, j);
  }
}

int next_greater(const vector<int> &a, int inv_pos) {
  for (int i = a.size() - 1; i > inv_pos; i--) {
    if (a[i] > a[inv_pos]) {
      return i;
    }
  }

  return -1;
}

bool next_permutation(vector<int> &a) {
  int inv_pos = -1;
  for (int i = a.size() - 1; i > 0; i--) {
    if (a[i] > a[i - 1]) {
      inv_pos = i - 1;
      break;
    }
  }

  if (inv_pos == -1) {
    return false;
  }

  int next_pos = next_greater(a, inv_pos);
  swap(a, inv_pos, next_pos);
  reverse(a, inv_pos + 1, a.size() - 1);

  return true;
}

// O(n) / O(1)
int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n;
  cin >> tt;

  while (tt--) {
    cin >> n;

    vector<int> a(n);
    for (int i = 0; i < n; i++) {
      cin >> a[i];
    }

    do {
      for (int e : a) {
        cout << e << " ";
      }
      cout << "\n";
    } while (next_permutation(a));
  }

  return 0;
}